src/libopensc/card-itacns.c

```diff
@@ -181,10 +191,11 @@ static int itacns_match_card(sc_card_t *card)
 	if(card->driver) {
 		DRVDATA(card)->ic_manufacturer_code = card->atr.value[i];
 		DRVDATA(card)->mask_manufacturer_code = card->atr.value[i+1];
-		DRVDATA(card)->os_version_h = card->atr.value[i+2];
-		DRVDATA(card)->os_version_l = card->atr.value[i+3];
 	}
-	i += 4; /* H3, H4, H5, H6 */
+	i += 2; /* H3, H4 */
+
+	card->version.fw_major = card->atr.value[i++] & 0xff; /* H5 */
+	card->version.fw_minor = card->atr.value[i++] & 0xff; /* H6 */
```

Comment:
The length checking here is not very obvious. Only by calling the very confusing macros `itacns_atr_match` (`itacns_atr_l`) the code checks whether the index is out of range or not. Unfortunately, the length is not checked when the code accesses other indices of the ATR.

In particular, you're accessing H3/H4 without checking if the ATR is long enough. 

Please take the time to review the whole initialization whether the length of the buffer is actually checked before accessing the bytes, thank you.

Reply:
Ok, I've moved things a bit around, given that some checks are depending on a variable index on the ATR, I had to use some utility functions, but now the initialization is split and no driver data is filled earlier than we've actually done the matching.